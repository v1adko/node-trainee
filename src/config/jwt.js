module.exports = {
  tokenSecret: process.env.SECRET_TOKEN_WORD || 'testSecretWord',
  tokenExpiresIn: '10h'
};
