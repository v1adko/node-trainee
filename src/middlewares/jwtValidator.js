import jwtService from '../services/jwtService';
import { TokenValidationError } from '../errors';

async function jwtValidator(request, response, next) {
  const token = request.headers['x-access-token'];
  if (!token) {
    return next();
  }

  try {
    const tokenData = await jwtService.decoder(token);
    request.user = tokenData;
  } catch (error) {
    next(error);
  }

  return next();
}

export default jwtValidator;
