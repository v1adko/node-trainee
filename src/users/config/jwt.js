module.exports = {
  secretTokkenWord: process.env.secretTokkenWord || 'secretword'
};
