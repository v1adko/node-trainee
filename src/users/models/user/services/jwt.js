const jwt = require('jsonwebtoken');

const { secretTokkenWord: secret } = require('../../../config/jwt');

const permissions = require('../../../config/permissions');

function generateJwt() {
  const payload = {
    _id: this._id,
    permissions: Object.values(permissions) // Gives all permissions
  };
  const options = {
    expiresIn: '2h'
  };

  return jwt.sign(payload, secret, options);
}

module.exports = {
  generateJwt
};

