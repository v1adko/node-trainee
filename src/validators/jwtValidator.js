import jwtService from '../services/jwtService';

async function verifyToken(request, response, next) {
  const token = request.headers['x-access-token'];
  if (!token) {
    return response
      .status(403)
      .send({ auth: false, message: 'No token provided.' });
  }

  try {
    const decodedToken = await jwtService.decoder(token);
    request.user = {
      id: decodedToken._id,
      role: decodedToken.role
    };
    next();
  } catch (error) {
    response.status(500).send({ auth: false, message: error.message });
  }

  return null;
}

export default verifyToken;
