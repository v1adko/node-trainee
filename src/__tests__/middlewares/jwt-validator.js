import jwt from 'jsonwebtoken';
import jwtValidator from '../../middlewares/jwt-validator';
import configJwt from '../../config/jwt';

const { tokenSecret: secret, tokenExpiresIn: expiresIn } = configJwt;
const options = { expiresIn };

let counter = 0;

const request = { headers: {} };

describe('Test jwt validator', () => {
  beforeEach(() => {
    counter = 0;
  });

  it('should not write in request user field, because token is missing', async () => {
    await jwtValidator(request, null, () => {
      counter += 1;
    });

    expect(counter).toBe(1);
    expect(request.user).toBeUndefined();
  });

  it('should pass in "next" error, because token is invalid', async () => {
    request.headers.access_token = 'token';

    const validationWithError = () =>
      jwtValidator(request, null, (error) => {
        throw error;
      });

    expect(request.user).toBeUndefined();
    expect(validationWithError()).rejects.toMatchSnapshot();
  });

  it('should return user data', async () => {
    const payload = {
      testField: 'test id'
    };
    const token = jwt.sign(payload, secret, options);

    request.headers.access_token = token;

    await jwtValidator(request, null, () => {
      counter += 1;
    });

    expect(counter).toBe(1);
    expect(request.user.testField).toBe(payload.testField);
  });
});
