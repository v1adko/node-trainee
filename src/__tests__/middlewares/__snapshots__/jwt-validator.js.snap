// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test jwt validator should pass in "next" error, because token is invalid 1`] = `[TokenValidationError: Invalid token, please repeat authentication.]`;
